src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs

```diff
@@ -129,58 +135,87 @@ public async ValueTask SynchronizeProjectAssetsAsync(ProjectStateChecksums proje
+
+            // Make parallel requests for all the project data across all projects at once.
+            tasks.Add(SynchronizeProjectAssetAsync<ProjectInfo.ProjectAttributes>(static p => p.Info, cancellationToken));
+            tasks.Add(SynchronizeProjectAssetAsync<CompilationOptions>(static p => p.CompilationOptions, cancellationToken));
+            tasks.Add(SynchronizeProjectAssetAsync<ParseOptions>(static p => p.ParseOptions, cancellationToken));
+            tasks.Add(SynchronizeProjectAssetCollectionAsync<ProjectReference>(static p => p.ProjectReferences, cancellationToken));
+            tasks.Add(SynchronizeProjectAssetCollectionAsync<MetadataReference>(static p => p.MetadataReferences, cancellationToken));
+            tasks.Add(SynchronizeProjectAssetCollectionAsync<AnalyzerReference>(static p => p.AnalyzerReferences, cancellationToken));
+
+            // Then sync each project's documents in parallel with each other.
+            foreach (var projectChecksums in allProjectChecksums)
+                tasks.Add(SynchronizeProjectDocumentsAsync(projectChecksums, cancellationToken));
+
+            await Task.WhenAll(tasks).ConfigureAwait(false);
+        }
+
+        static void AddAll(HashSet<Checksum> checksums, ChecksumCollection checksumCollection)
+        {
+            foreach (var checksum in checksumCollection)
+                checksums.Add(checksum);
+        }
+
+        async Task SynchronizeProjectAssetAsync<TAsset>(Func<ProjectStateChecksums, Checksum> getChecksum, CancellationToken cancellationToken)
+        {
+            await Task.Yield();
             using var _ = PooledHashSet<Checksum>.GetInstance(out var checksums);
 
-            checksums.Add(projectChecksums.Info);
-            checksums.Add(projectChecksums.CompilationOptions);
-            checksums.Add(projectChecksums.ParseOptions);
-            AddAll(checksums, projectChecksums.ProjectReferences);
-            AddAll(checksums, projectChecksums.MetadataReferences);
-            AddAll(checksums, projectChecksums.AnalyzerReferences);
-            AddAll(checksums, projectChecksums.Documents.Checksums);
-            AddAll(checksums, projectChecksums.AdditionalDocuments.Checksums);
-            AddAll(checksums, projectChecksums.AnalyzerConfigDocuments.Checksums);
+            foreach (var projectChecksums in allProjectChecksums)
+                checksums.Add(getChecksum(projectChecksums));
 
-            // First synchronize all the top-level info about this project.
-            await this.SynchronizeAssetsAsync<object, VoidResult>(
-                assetPath: AssetPath.ProjectAndDocuments(projectChecksums.ProjectId), checksums, callback: null, arg: default, cancellationToken).ConfigureAwait(false);
+            await this.SynchronizeAssetsAsync<TAsset, VoidResult>(
+                AssetPath.SolutionAndTopLevelProjectsOnly, checksums, callback: null, arg: default, cancellationToken).ConfigureAwait(false);
```

Comment:
>AssetPath.SolutionAndTopLevelProjectsOnly [](http://example.com/codeflow?start=16&length=41)

Any concern that for a small number of projects being inspected, that we are now looking through all projects for their assets? Generally, it feels that for a small number of projects being synchronized that the earlier approach would have worked better.

Reply:
let me noodle on that.  perhaps a blended approach will be better.  when it's a small number, just do a project at a time.  when we're doing a lot, do the bulk search.

Reply:
moved to the hybrid approach.  direct project-sync when we have a small number.  bulk when we have a large number.